compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1090,9 +1107,15 @@ impl<'hir> LoweringContext<'_, 'hir> {
         coroutine_kind: Option<CoroutineKind>,
         body: Option<&Block>,
     ) -> hir::BodyId {
-        let (Some(coroutine_kind), Some(body)) = (coroutine_kind, body) else {
+        let Some(coroutine_kind) = coroutine_kind else {
+            // Normal case: not a coroutine.
             return self.lower_fn_body_block(span, decl, body);
         };
+        let Some(body) = body else {
+            // Coroutine without a body... always an error.
+            return self
+                .lower_fn_body(decl, |this| this.expr_err(span, this.dcx().has_errors().unwrap()));
+        };
```

Comment:
It seems like the infinite loop can leak into error messages via RPIT, damn... we should probably only do this fallback here for intrinsics, but I have not yet figured out how to access attributes here. Given that attributes are available even earlier during AST validation I hope it's not going to be too hard.